AsyncNats/Messages/NatsMessageParser.cs

```diff
@@ -2,80 +2,407 @@
+
+                //get pointers to strings
+                NatsKey subject;
+                NatsKey replyTo = NatsKey.Empty;
+
+
+                copyMemory = copyRented.Memory;
+                if (splitCount == 1)
+                {
+                    subject = new NatsKey(copyMemory.Slice(4, splits[0] - 4));
+                }
+                else
+                {
+                    replyTo = new NatsKey(copyMemory.Slice(splits[0] + 1, payloadSizeStart - splits[0] - 1));
+
+                    subject = new NatsKey(copyMemory.Slice(4, splits[1] - 4));
+                }
+
+                return new NatsMsg(subject, sid, replyTo, payload, ReadOnlyMemory<byte>.Empty, copyRented);
+
+            }
+
+
+
+
+        }
+
+        public NatsMsg? ParseMessageWithHeader(in ReadOnlySpan<byte> line, ref SequenceReader<byte> reader, out bool inlined)
+        {
+            inlined = false;
+            
+            //parse total size
+            var multiplier = 1;
+            var totalSize = 0;
+            var pointer = line.Length - 1;
+
+            ref byte lineRef = ref MemoryMarshal.GetReference(line);
+            byte currentByte = Unsafe.Add(ref lineRef, pointer);
+            do
+            {
+                totalSize += (currentByte - '0') * multiplier;
+                multiplier *= 10;
+                pointer--;
+                currentByte = Unsafe.Add(ref lineRef, pointer);
+            } while (currentByte != ' ');
+
+            if (reader.Remaining < totalSize + 2) return null;
+
+            pointer--;
+
+            //parse header size
+            multiplier = 1;
+            var headerSize = 0;
+            var headerSizeStart = pointer;
+            currentByte = Unsafe.Add(ref lineRef, pointer);
+            do
+            {
+                headerSize += (currentByte - '0') * multiplier;
+                multiplier *= 10;
+                pointer--;
+                currentByte = Unsafe.Add(ref lineRef, pointer);
+            } while (currentByte != ' ');
+
+
+            pointer--;
+            var headerSizeEnd = pointer;
+
+            Span<int> splits = stackalloc int[3];
+            var splitCount = 0;
+
+            while (pointer > 4)
+            {
+                var ch = line[pointer];
+                if (ch == ' ')
+                {
+                    splits[splitCount] = pointer;
+                    splitCount++;
+
+                    if (splitCount > 2)
+                        throw new ProtocolViolationException($"Invalid message header {Encoding.UTF8.GetString(line)}");
+                }
+
+                pointer--;
+            }
+            var payloadSize = totalSize - headerSize;
+
+            
+
+            //sid
+            long sid = 0;
+            pointer = (splitCount == 1) ? headerSizeEnd : splits[0] - 1;
+            multiplier = 1;
+            currentByte = Unsafe.Add(ref lineRef, pointer);
```

Comment:
This is faster than Utf8Parser.TryParse?

Reply:
It is in this particular case because digit count is known and we assume bytes are in the right format, avoiding some bound checks.